src/package.rs

```diff
@@ -0,0 +1,519 @@
+  fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+    Some(self.cmp(other))
+  }
+}
+
+// Sort the package requirements alphabetically then the version
+// requirement in a way that will lead to the least number of
+// duplicate packages (so sort None last since it's `*`), but
+// mostly to create some determinism around how these are resolved.
+impl Ord for PackageReq {
+  fn cmp(&self, other: &Self) -> Ordering {
+    fn cmp_specifier_version_req(a: &VersionReq, b: &VersionReq) -> Ordering {
+      match a.tag() {
+        Some(a_tag) => match b.tag() {
+          Some(b_tag) => b_tag.cmp(a_tag), // sort descending
+          None => Ordering::Less,          // prefer a since tag
+        },
+        None => {
+          match b.tag() {
+            Some(_) => Ordering::Greater, // prefer b since tag
+            None => {
+              // At this point, just sort by text descending.
+              // We could maybe be a bit smarter here in the future.
+              b.to_string().cmp(&a.to_string())
+            }
+          }
+        }
+      }
+    }
+
+    match self.name.cmp(&other.name) {
+      Ordering::Equal => {
+        cmp_specifier_version_req(&self.version_req, &other.version_req)
+      }
+      ordering => ordering,
+    }
+  }
+}
+
+#[derive(Debug, Error, Clone)]
+#[error("Invalid package name and version reference '{text}'. {message}")]
+pub struct PackageNvReferenceParseError {
+  pub message: String,
+  pub text: String,
+}
+
+/// A package name and version with a potential subpath.
+#[derive(Debug, Clone, PartialOrd, Ord, PartialEq, Eq, Hash)]
+pub struct PackageNvReference {
+  pub nv: PackageNv,
+  pub sub_path: Option<String>,
+}
+
+impl PackageNvReference {
+  #[allow(clippy::should_implement_trait)]
+  pub(crate) fn from_str(
+    nv: &str,
+    kind: PackageKind,
+  ) -> Result<Self, PackageNvReferenceParseError> {
+    use monch::*;
+
+    fn sub_path(input: &str) -> ParseResult<&str> {
+      let (input, _) = ch('/')(input)?;
+      Ok(("", input))
+    }
+
+    fn parse_ref<'a>(
+      kind: PackageKind,
+    ) -> impl Fn(&'a str) -> ParseResult<'a, PackageNvReference> {
+      move |input| {
+        let (input, _) = tag(kind.scheme_with_colon())(input)?;
+        let (input, nv) = parse_nv(input)?;
+        let (input, maybe_sub_path) = maybe(sub_path)(input)?;
+        Ok((
+          input,
+          PackageNvReference {
+            nv,
+            sub_path: maybe_sub_path.map(ToOwned::to_owned),
+          },
+        ))
+      }
+    }
+
+    with_failure_handling(parse_ref(kind))(nv).map_err(|err| {
+      PackageNvReferenceParseError {
+        message: format!("{err:#}"),
+        text: nv.to_string(),
+      }
+    })
+  }
+
+  pub(crate) fn as_specifier(&self, kind: PackageKind) -> Url {
+    let version_text = self.nv.version.to_string();
+    let scheme = kind.scheme();
+    let capacity = scheme.len() + 2 /* colon slash */
+      + self.nv.name.len()
+      + 1 /* @ */
```

Comment:
Is this correct if the version is "*"?

Reply:
The version will always be resolved here, so it will be something like x.x.x